internal/querycoordv2/job/job.go

```diff
@@ -0,0 +1,487 @@
+package job
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/milvus-io/milvus/internal/log"
+	"github.com/milvus-io/milvus/internal/proto/querypb"
+	"github.com/milvus-io/milvus/internal/querycoordv2/meta"
+	"github.com/milvus-io/milvus/internal/querycoordv2/session"
+	"github.com/milvus-io/milvus/internal/querycoordv2/utils"
+	"github.com/milvus-io/milvus/internal/util/typeutil"
+	"github.com/samber/lo"
+	"go.uber.org/zap"
+)
+
+// Job is request of loading/releasing collection/partitions,
+// the execution flow is:
+// 1. PreExecute()
+// 2. Execute(), skip this step if PreExecute() failed
+// 3. PostExecute()
+type Job interface {
+	MsgID() int64
+	CollectionID() int64
+	// PreExecute does checks, DO NOT persists any thing within this stage,
+	PreExecute() error
+	// Execute processes the request
+	Execute() error
+	// PostExecute clears resources, it will be always processed
+	PostExecute()
+	Error() error
+	SetError(err error)
+	Done()
+	Wait() error
+}
+
+type BaseJob struct {
+	ctx          context.Context
+	msgID        int64
+	collectionID int64
+	err          error
+	doneCh       chan struct{}
+}
+
+func NewBaseJob(ctx context.Context, msgID, collectionID int64) *BaseJob {
+	return &BaseJob{
+		ctx:          ctx,
+		msgID:        msgID,
+		collectionID: collectionID,
+		doneCh:       make(chan struct{}),
+	}
+}
+
+func (job *BaseJob) MsgID() int64 {
+	return job.msgID
+}
+
+func (job *BaseJob) CollectionID() int64 {
+	return job.collectionID
+}
+
+func (job *BaseJob) Error() error {
+	return job.err
+}
+
+func (job *BaseJob) SetError(err error) {
+	job.err = err
+}
+
+func (job *BaseJob) Done() {
+	close(job.doneCh)
+}
+
+func (job *BaseJob) Wait() error {
+	<-job.doneCh
+	return job.err
+}
+
+func (job *BaseJob) PreExecute() error {
+	return nil
+}
+
+func (job *BaseJob) PostExecute() {}
+
+type LoadCollectionJob struct {
+	*BaseJob
+	req *querypb.LoadCollectionRequest
+
+	dist      *meta.DistributionManager
+	meta      *meta.Meta
+	targetMgr *meta.TargetManager
+	broker    meta.Broker
+	nodeMgr   *session.NodeManager
+}
+
+func NewLoadCollectionJob(
+	ctx context.Context,
+	req *querypb.LoadCollectionRequest,
+	dist *meta.DistributionManager,
+	meta *meta.Meta,
+	targetMgr *meta.TargetManager,
+	broker meta.Broker,
+	nodeMgr *session.NodeManager,
+) *LoadCollectionJob {
+	return &LoadCollectionJob{
+		BaseJob:   NewBaseJob(ctx, req.Base.GetMsgID(), req.GetCollectionID()),
+		req:       req,
+		dist:      dist,
+		meta:      meta,
+		targetMgr: targetMgr,
+		broker:    broker,
+		nodeMgr:   nodeMgr,
+	}
+}
+
+func (job *LoadCollectionJob) PreExecute() error {
+	req := job.req
+	log := log.With(
+		zap.Int64("msgID", req.Base.GetMsgID()),
+		zap.Int64("collection", req.GetCollectionID()),
+	)
+
+	if req.GetReplicaNumber() <= 0 {
```

Comment:
Why is replica number less than 0?

Reply:
old SDK may send a request without ReplicaNumber, which leads to 0